src/avt/Database/Ghost/avtUnstructuredDomainBoundaries.C

```diff
@@ -600,29 +600,38 @@ avtUnstructuredDomainBoundaries::ExchangeMeshT(vector<int>       domainNum,
 //    Eric Brugger, Fri Mar 13 15:20:08 PDT 2020
 //    Modify to handle NULL meshes.
 //
+//    Kathleen Biagas, Thu Oct 31, 2024
+//    Ensure all procs are calling the same Exchange function.
+//
 // ****************************************************************************
 
 vector<vtkDataArray*>
 avtUnstructuredDomainBoundaries::ExchangeScalar(vector<int>         domainNum,
                                               bool                isPointData,
                                               vector<vtkDataArray*> scalars)
 {
-    // We're in a bit of a sticky situation if we don't have any actual data.
-    // Without a valid vtkDataArray, we don't know which ExchangeData to
-    // call. But if there's no data, nothing will actually be exchanged
-    // aside from basic communications (eg: domain2proc, MPI_Barrier),
-    // so we'll choose to call one. 
-    if (!scalars.size())
+    int nonNullDomain = 0;
+    int dataType = -1;
+    if(!scalars.empty())
     {
-        return ExchangeData_float(domainNum, isPointData, scalars);
+        while (nonNullDomain < domainNum.size() && scalars[nonNullDomain] == NULL)
+            nonNullDomain++;
+        dataType = scalars[nonNullDomain]->GetDataType();
     }
+
+#ifdef PARALLEL
+    // Let's get them all to agree on one data type.
+    int myDataType = dataType;
+    MPI_Allreduce(&myDataType, &dataType, 1, MPI_INT, MPI_MAX, VISIT_MPI_COMM);
+#endif
+
+    if (dataType < 0)
```

Comment:
So, I am wondering if this is still a bit off. I mean, I think it will work ok but if the reason for introducing this block was that all processors did NOT agree on the value of `dataType`, after making this call, all processors will now agree on whatever the *maximum* value was among them. Certainly, if that maximum value is `< 0`, then no processor has data to contribute here and we can short-circuit.

However, what if that maximum value is `>=0` but nonetheless not equal to the value for `dataType` obtained on 619? That could happen if for some reason two processors had data to contribute but the types were different. Now, I suspect that is truly highly unlikely. But, its also easily accounted for by a slight swizzle on what is here...

```
    int maxDataType = dataType;
ifdef PARALLEL
     MPI_Allreduce(&dataType, &maxDataType, 1, MPI_INT, MPI_MAX, VISIT_MPI_COMM);
 #endif

     if (maxDataType < 0 || maxDataType != dataType)
```

Or, maybe you'd want to decouple these two cases as one is more or less an error and the other is potentially a common occurence for the case that no processors have data.

Thoughts?

Reply:
I agree the logic looks completely flawed. It assumes that all the vtkDataArrays stored in the input vector 'scalars' have the same data type.  This is true in the similar functions of avtStructuredDomainBoundaries.

I took a look at all the places where ExchangeScalar is called and I see:

Retrieval of array from either point or cell data for each domain. It is either the generic Scalars array, or a named array, and these would all have the same data type.

Retrieval of extraGhostNodes/Zones array form each domain. These would all have the same data type.

I discovered similar for the calls to ExchangeVector.

So the ExchangeScalar/Vector methods are called with a list of the same vtkDataArray from different domains, so the assumption that all the vtkDataArrays in the input vector all have the same data type is valid.

The only problem lay in the fact that this method used a default 'FLOAT' exchange for a processor that had no data.

Reply:
While there is nothing really I do not wholeheartedly agree with in what you've explained, the code as is will be guaranteed to work when the assumptions regarding types are true (which they almost certainly always will be) whereas the very, very slight modification I propose will ensure it works whether or not those assumptions hold. I might also add, they avert anyone else arriving here in this block of code (perhaps without being aware of those assumptions) and wondering, as I did, about that case.

Reply:
>>Or, maybe you'd want to decouple these two cases as one is more or less an >>error and the other is potentially a common occurence for the case that no >>processors have data.

What would you suggest as the best handling for the case where multiple procs have data but the dataTypes don't match? Thrown an Exception? Write a warning to the debug log? Issue a visible warning  (eg popup in the gui) , if its possible from here.

Reply:
And should I add the same logic to similar code in avtStructuredDomainBoundaries?

Reply:
I would.

Reply:
> What would you suggest as the best handling for the case where multiple procs have data but the dataTypes don't match? Thrown an Exception?

Here are my thoughts...

* Crashing is worst...thats what this PR is fixing 💪🏻 
* Failing silently so that nobody knows why is better but unsatisfactory due to lack of knowing cause if issue
* Excepting with a message is best but it may not be practical

IMHO, there is a fate worse than crashing...parallel deadlock. Its as bad as crashing because it is irrecoverable and progress will not proceed further. It is worse than crashing becasue nobody knows it's happened...until you have maybe waited longer than your threshold of tolerance.